File: day07/ssl.pl
#!/usr/bin/env perl
use strict;
use warnings;
use 5.24.0;
use feature 'signatures';
no warnings "experimental::signatures";

my $n = 0;

while (<>) {
    chomp;
    my $orig = $_;
    say $orig;

    # assume [ and ] come in pairs and aren't nested
#    $_ = "x$_" if substr($_, 0, 1) eq "[";
    s/[\[\]]/ /g;
    my @s = split / /;

    my @aba;
    my @bab;
    for my $i (0..$#s) {
        if ($i % 2 == 0) {
            push @aba, @{find_aba($s[$i])};
        } else {
            push @bab, @{find_aba($s[$i])};
        }
    }

    if (has_ssl(\@aba, \@bab)) {
        say "GOOD => $orig";
        $n++;
    } else {
        say "BAD  => $orig";
    }


}

say $n;

sub find_aba($s) {
    my @c = split //, $s;
    my @a;

    for my $i (0..$#c - 2) {
        if ($c[$i] eq $c[$i+2] && $c[$i] ne $c[$i+1]) {
            push @a, substr($s, $i, 3);
        }
    }
    return \@a;
}

sub has_ssl($aba, $bab) {
    for my $s (@$aba) {
        my @c = split //, $s;
        my $rev = sprintf "^%s%s%s\$", $c[1], $c[0], $c[1];
        return 1 if grep /$rev/, @$bab;
    }

    return 0;
}

File: day07/ssl.py
#!/usr/bin/env python
from sys import argv
import re

def find_aba(s) :
    c = list(s)
    a = []

    for i in range(0, len(c) - 2) :
        if c[i] == c[i+2] and c[i] != c[i+1] :
            a.append(s[i:i+3])

    return a

def has_ssl(aba, bab) :
    
    for s in aba :
        c = list(s)
        rev = c[1] + c[0] + c[1]
        for t in bab :
            if t == rev :
                return True

    return False

n = 0

script, filename = argv
with open(filename) as f :
    for line in f :
        line = line.rstrip("\n")

        tok = re.split('[\[\]]', line)

        aba = []
        bab = []
        for i in range(0, len(tok)) :
            if i % 2 == 0 :
                abas = find_aba(tok[i])
                if len(abas) > 0 :
                    aba = aba + abas

            else :
                abas = find_aba(tok[i])
                if len(abas) > 0 :
                    bab = bab + abas

        if has_ssl(aba, bab) :
            print "GOOD => %s" % line
            n = n + 1
        else :
            print "BAD  => %s" % line

print n
